src/concrete_intersection.jl

```diff
@@ -176,6 +174,148 @@ function intersection(x::Interval{N},
+- `hs` -- half-space
+
+### Output
+
+If the sets do not intersect, the result is the empty set.
+If the interval is fully contained in the half-space, the result is the original
+interval.
+Otherwise the result is the interval that describes the intersection.
+"""
+function intersection(X::Interval{N}, hs::HalfSpace{N}
+                     )::Union{Interval{N}, EmptySet{N}} where {N<:Real}
+    @assert dim(hs) == 1 "cannot take the intersection between an interval " *
+                         "and a $(dim(hs))-dimensional half-space"
+
+    a = hs.a[1]
+    b = hs.b
+    is_leq = a >= zero(N)
+
+    close_bound = is_leq ? min(X) : max(X)
+    if !_leq(close_bound * a, b)
+        # sets are disjoint
+        return EmptySet{N}()
+    end
+
+    far_bound = is_leq ? max(X) : min(X)
+    if _leq(far_bound * a, b)
+        # interval is fully contained
+        return X
+    end
+
+    # proper intersection; compute new interval
+    new_bound = b / a
+    if is_leq
+        Y = Interval(close_bound, new_bound)
+    else
+        Y = Interval(new_bound, close_bound)
+    end
+    return Y
+end
+
+# symmetric method
+function intersection(hs::HalfSpace{N}, X::Interval{N}
+                     )::Union{Interval{N}, EmptySet{N}} where {N<:Real}
+    return intersection(X, hs)
+end
+
+"""
+    intersection(X::Interval{N}, hp::Hyperplane{N}
+                )::Union{Singleton{N}, EmptySet{N}} where {N<:Real}
+
+Compute the intersection of an interval and a hyperplane.
+
+### Input
+
+- `X`  -- interval
+- `hp` -- hyperplane
+
+### Output
+
+If the sets do not intersect, the result is the empty set.
+Otherwise the result is the singleton that describes the intersection.
+"""
+function intersection(X::Interval{N}, hp::Hyperplane{N}
+                     )::Union{Singleton{N}, EmptySet{N}} where {N<:Real}
+    @assert dim(hp) == 1 "cannot take the intersection between an interval " *
+                         "and a $(dim(hp))-dimensional hyperplane"
+
+    # a one-dimensional hyperplane is just a point
+    p = hp.b / hp.a[1]
+    if _leq(min(X), p) && _leq(p, max(X))
+        return Singleton([p])
+    else
+        return EmptySet{N}()
+    end
+end
+
+# symmetric method
+function intersection(hp::Hyperplane{N}, X::Interval{N}
+                     )::Union{Singleton{N}, EmptySet{N}} where {N<:Real}
+    return intersection(X, hp)
+end
+
+"""
+    intersection(X::Interval{N}, Y::LazySet{N}
+                )::Union{Interval{N}, EmptySet{N}} where {N<:Real}
+
+Compute the intersection of an interval and a convex set.
+
+### Input
+
+- `X` -- interval
+- `Y` -- convex set
+
+### Output
+
+If the sets do not intersect, the result is the empty set.
+Otherwise the result is the interval that describes the intersection.
+"""
+function intersection(X::Interval{N}, Y::LazySet{N}
+                     )::Union{Interval{N}, EmptySet{N}} where {N<:Real}
+    lower = max(min(X), -ρ(N[-1], Y))
+    upper = min(max(X), ρ(N[1], Y))
+    if lower > upper
```

Comment:
How about if we add robustness here too? What i have in mind is:

```julia
if lower < upper
    return Interval(lower, upper)
elseif _isapprox(lower, upper)
    return Singleton([lower])
else
    return EmptySet()
end
```

Otherwise, we can always return a one element interval (for which `isflat` could be true though, but it's not a big deal?),

```julia
if _leq(lower, upper)
    return Interval(lower, upper)
else
    return EmptySet()
end
```

Reply:
I do not like to create `Interval`s whose lower bound is above the upper bound, even by a small amount.

Reply:
Ok and how about the first paragraph? Now we got:

```julia
julia> X
Interval{Float64,IntervalArithmetic.Interval{Float64}}([1, 2])

julia> S = Singleton([1.0]);

julia> intersection(X, S ⊕ [1.0 + LazySets.ABSZTOL(Float64)])
EmptySet{Float64}()

julia> intersection(X, S ⊕ [1.0 + 1e-15])
EmptySet{Float64}()

julia> intersection(X, S ⊕ [1.0 + 1e-16])
Interval{Float64,IntervalArithmetic.Interval{Float64}}([2, 2])
```
I think we want that for all values smaller than `LazySets.ABSZTOL(Float64)` it returns the interval.

Reply:
`_isapprox` is only defined for `AbstractFloat`s. Is that on purpose?

Reply:
Yes, `_isapprox` was written for `AbstractFloat`, since that rationals can be compared exactly. The definition of [isapprox](https://github.com/JuliaLang/julia/blob/9a1cc63e7186b50cd080922105ba56b3d1b89b01/base/floatfuncs.jl#L273) in Julia uses `Number`. I agree that for methods that use `_isapprox` generically it make sense that it is defined for `Real`; this will use `isapproxzero(x::Real; kwargs...)` that uses the fallback `x == zero(x)` for everything which is not float.